File: Python/Timsort.py
"""
Timsort
If our list has less than 64 elements, run insertion sort!
This is a very dumped down version of the actual timsort.
Its way slower than the actual built in sort, but its good enough
for me to learn and understand how it works.

First we use insertionsort on Runs in our list.
Now we have multiple sorted runs.
Next we merge the runs and create bigger sorted runs.
Repeat until done.
Done
"""

"""
RUN is our size of sorted blocks (known as Run in timsort)
"""
RUN = 32


def insertion_sort(arr, start=0, end=None):
    if end is None:
        end = len(arr)
    for i in range(start, end):
        elem = arr[i]
        j = i
        while j > 0 and arr[j - 1] > elem:
            arr[j] = arr[j - 1]
            j -= 1
        arr[j] = elem
    return arr


def merge(arr, start, mid, end):
    # Slice our original array in 2 sorted subarrays
    len1, len2 = mid - start + 1, end - mid
    left, right = arr[start: mid + 1], arr[mid + 1: end + 1]

    i, j, k = 0, 0, start
    # Compare each element of the 2 sorted subarrays
    # and merge them back in the original array
    while i < len1 and j < len2:
        if left[i] <= right[j]:
            arr[k] = left[i]
            i += 1
        else:
            arr[k] = right[j]
            j += 1
        k += 1

    while i < len1:
        arr[k] = left[i]
        k += 1
        i += 1

    while j < len2:
        arr[k] = right[j]
        k += 1
        j += 1


def timsort(arr):
    n = len(arr)
    if n <= 64:  # On small lists its best to use insertion_sort to minimize function calls
        return insertion_sort(arr)
    else:
        for i in range(0, n, RUN):
            insertion_sort(arr, i, min((i + RUN - 1), (n - 1)))
        # At this point we have one unsorted list
        # containing multiple sorted runs/ blocks

        # Now we need to take 2 runs/ blocks and merge them
        # After each merge we double our merge size
        # that we merge our old merged subarray with the new subarray
        block = RUN
        while block < n:
            for start in range(0, n, 2 * block):
                end = min((start + 2 * block - 1), (n - 1))
                mid = (start + end) // 2
                merge(arr, start, mid, end)

            block *= 2


if __name__ == "__main__":
    test_arr = [
        3,
        1,
        7,
        4,
        12,
        9,
        11,
        102
    ]

    timsort(test_arr)
    print(test_arr)
    import benchmark
    benchmark.benchmark(timsort)

File: C++/Timsort.cpp
#include <vector>
#include <iostream>
#include <algorithm>  // used for min(x, y);

#include "Randomdata.cpp"

#define RUN 32

void insertionSort(std::vector<int> &arr, int start = 0, int end = 0){
    int n = arr.size();
    for (int i = 1; i < n; i++){
        int elem = arr[i];
        int j = i;
        while (j > 0 && arr[j - 1] > elem){
            arr[j] = arr[j - 1];
            j--;
        }
        arr[j] = elem;
    }
}

void merge(std::vector<int> &arr, int start, int mid, int end){
    int len1 = mid - start + 1;
    int len2 = end - mid;

    std::vector<int> left(len1);
    std::vector<int> right(len2);

    for (int i = 0; i < len1; i++){
        left[i] = arr[start + i];
    }
    for (int i = 0; i < len2; i++){
        right[i] = arr[mid + 1 + i];
    }

    int i = 0;
    int j = 0;
    int k = start;

    while (i < len1 && j < len2){
        if (left[i] <= right[j]){
            arr[k] = left[i];
            i++;
        }
        else{
            arr[k] = right[j];
            j++;
        }
        k++;
    }

    while (i < len1){
        arr[k] = left[i];
        k++;
        i++;
    }

    while (j < len2){
        arr[k] = right[j];
        k++;
        j++;
    }

}

void timSort(std::vector<int> &arr){
    int n = arr.size();
    if (n <= 64){
        return insertionSort(arr, 0, n - 1);
    }

    for (int i = 0; i < n; i += RUN){
        insertionSort(arr, i, std::min(i + RUN - 1, n - 1));
    }

    int block = RUN;
    while (block < n){
        int start, end, mid;
        for (start = 0; start < n; start += 2 * block){
            end = std::min(n - 1, start + 2 * block - 1);
            mid = (start + end) / 2;
            merge(arr, start, mid, end);
        }
        block = block * 2;
    }
}

void printArray(const std::vector<int>& arr){
    for (auto const& v: arr){
        std::cout << v << " | ";
    }
}

int main(){
    std::vector<int> arr = {12, 321, 12, 1, 5, 34, 27};
    timSort(arr);
    printArray(arr);

    std::cout << "\n\n";
    std::vector<int> bench = randomData(5000);
    printArray(bench);
    std::cout << "\n\n";
    timSort(bench);
    printArray(bench);

    return 0;
}